docs/user-guide/src/tour/loop_invariants.md

```diff
@@ -0,0 +1,25 @@
+# Loop Invariants
+
+To show how to verify loops, we will use a different example than our linked list for simplicity.
+We will write and verify a function that can add some value to every element of an array slice.
+
+Lets write a function that takes an integer `x` and sums up all values from 0 to that value in a loop.
+For non-negative inputs, the result will be equal to `x * (x + 1) / 2`:
+
+```rust,noplaypen
+{{#rustdoc_include ../../../../prusti-tests/tests/verify/fail/user-guide/loop_invariants.rs:specification}}
+// Prusti: fails
+```
+
+We cannot verified this code yet, because Prusti does not know what the `while` loop does to `sum` and `i`. For that, we need to add a [`body_invariant`](../verify/loop.md). Body invariants are expressions that always hold at the beginning and end of the loop body. In our case, the invariant is that `sum` contains the sum of all values between 1 and `i`. Since `i` starts at 1 and not at 0, we have to slightly adjust the formula by using `i - 1` instead of `i`, so we get: `sum == (i - 1) * i / 2`.
```

Comment:
"cannot verify"

Also the part "always hold at the beginning and end of the loop body" is not exactly true. The invariant is assumed or checked in the place it is written in the code (it could be in the middle of a loop). Then:

- When entering the loop on the first iteration, the invariant must hold when reached (which implies the loop condition was true);
- when re-entering the loop on an arbitrary iteration, the invariant is assumed, then, executing the rest of the loop and the next iteration's beginning (until the invariant), the invariant must hold again;
- when exiting the loop, the invariant is assumed.

… which can sound rather confusing without a good diagram and should rather be explained in the body invariant section proper.